src/Microsoft.ML.TimeSeries/SeasonalityDetector.cs

```diff
@@ -0,0 +1,317 @@
+            var periodogram = energies.Select((t, i) => t * Complex.Conjugate(t)).ToArray();
+            FindBestTwoFrequencies(periodogram, length, out var bestFreq, out var secondFreq);
+
+            double[] ifftRe = new double[length];
+            double[] ifftIm = new double[length];
+            FftUtils.ComputeBackwardFft(
+                periodogram.Select(t => t.Real).ToArray(),
+                periodogram.Select(t => t.Imaginary).ToArray(),
+                ifftRe,
+                ifftIm,
+                length);
+            var values = ifftRe.Select((t, i) => new Complex(t, ifftIm[i])).ToArray();
+
+            int period = FindActualPeriod(values, bestFreq, secondFreq, length, randomessThreshold);
+            return period < 0 ? -1 : period;
+        }
+
+        /// <summary>
+        /// Find the actual period based on best frequency and second best frequency:
+        /// Pick the best frequency by inspecting the auto-correlation energy (pick the highest) in time-domain.
+        /// In the normal case, usually, when the time series is with period T, then the best frequency is N/T,
+        /// while the second frequency would be N/2T, because period = T implies period = nT, where n is an integer.
+        /// In such a case, smaller period will win out on the autu-correlation energy list, due to the property
+        /// of auto-correlation.
+        /// </summary>
+        /// <param name="values">The auto-correlation function of the augmented time series</param>
+        /// <param name="bestFrequency">The best frequency candidate</param>
+        /// <param name="secondFrequency">The second best frequency candidate</param>
+        /// <param name="timeSeriesLength">The length of the original time series, this is used for post processing to reduce false positive</param>
+        /// <param name="randomnessThreshold">Randomness flutuation threashold  that specifies how confidence the input
+        /// follows a predictable pattern recurring over an interval as seasonal data.
+        /// </param>
+        /// <returns>The period detected by check best frequency and second best frequency</returns>
+        private static int FindActualPeriod(Complex[] values, int bestFrequency, int secondFrequency, int timeSeriesLength, double randomnessThreshold)
+        {
+            int firstPeriod = -1;
+            int secondPeriod = -1;
+            double firstTimeDomainEnergy = -1;
+            double secondTimeDomainEnergy = -1;
+            firstPeriod = FindBestPeriod(values, bestFrequency, timeSeriesLength, out firstTimeDomainEnergy);
+            if (secondFrequency != -1)
+            {
+                secondPeriod = FindBestPeriod(values, secondFrequency, timeSeriesLength, out secondTimeDomainEnergy);
+            }
+            if (firstPeriod == -1 && secondPeriod == -1)
+                return -1;
+            int truePeriod;
+            double trueTimeDomainEnergy;
+            if (firstPeriod == -1)
```

Comment:
nit: Please add some spacing between blocks for readability.

Reply:
addressed